### ISSUE TITLE
Event not invoked in ShadowDom's non-root node.

### ISSUE THREAD

**Issue by Jack-Works** [facebook/react#15759]
**Do you want to request a *feature* or report a *bug*?**
Bug
**What is the current behavior?**
The second h1 onClick not responding

**Reproduce:**
```html
<script crossorigin src="https://unpkg.com/react@16/umd/react.development.js"></script>
<script crossorigin src="https://unpkg.com/react-dom@16/umd/react-dom.development.js"></script>

<body>
    <div id="1"></div>
    <div id="2"></div>
</body>
<script>
    function Component() {
        return React.createElement('h1', {onClick() { alert('click!') }}, 'Click me!')
    }
    const shadow1 = document.getElementById('1').attachShadow({mode: 'closed'})
    const shadow2 = document.getElementById('2').attachShadow({mode: 'closed'})
    const inner = document.createElement('div')
    shadow2.appendChild(inner)

    ReactDOM.render(React.createElement(Component), shadow1)
    ReactDOM.render(React.createElement(Component), inner)
</script>
```

**What is the expected behavior?**
The first h1 works, the second h1 works too.

**Comment by Jack-Works** [comment]
If anyone interested, here is a hack that can resolve problems by event lost due to ShadowRoot.
Full example at https://github.com/mui-org/material-ui/issues/16223

```javascript
{
        // ? Hack for React, let event go through ShadowDom
        const hackingEvents = new WeakMap()
        function hack(eventName, shadowRoot) {
            shadowRoot.addEventListener(eventName, e => {
                if (hackingEvents.has(e)) return
                const path = e.composedPath()
                var e2 = new e.constructor(e.type, e)
                hackingEvents.set(e2, path)
                portal.dispatchEvent(e2)
                e.stopPropagation()
                e.stopImmediatePropagation()
            })
        }
        // call hack with parameter1 = Event name you want to make it work with ShadowRoot
        // parameter2 = the shadow root
        hack('click', portal)
        var nativeTarget = Object.getOwnPropertyDescriptor(Event.prototype, 'target').get
        Object.defineProperty(Event.prototype, 'target', {
            get() {
                if (hackingEvents.has(this)) return hackingEvents.get(this)[0]
                return nativeTarget.call(this)
            }
        })
    }
```

**Comment by Jack-Works** [comment]
Wait. This bug is not fixed yet. Don't close it.

**Comment by Jack-Works** [comment]
Don't close

**Comment by Jack-Works** [comment]
Bump

**Comment by Jack-Works** [comment]
Let me try if it is fixed in react 17 later

**Comment by Jack-Works** [comment]
resolved in 17.0.0-rc.3
